### PR TITLE
PERF: Implement get_freq_code in cython frequencies

### PR DESCRIPTION
**Pull request by jbrockmendel**
There's a whole bunch of `tseries.frequencies` and `tseries.offsets` that should be moved into cython to improve `Period` (and `PeriodIndex, and to a lesser extend `Timestamp`) performance.  This PR just starts with the function `get_freq_code`.

To keep the PR small, there are a bunch of things it _doesnt_ do that can further improve things in follow-ups:
 - Nothing is cdef'd or declared in a pxd file
 - `_libs.period` still gets `get_freq_code` from `tseries.frequencies` instead of directly from `tslibs.frequencies`
 - A handful of now-duplicate constants in `tseries.frequencies` are not removed.

```
In [6]: per = pd.Period.now('min')
In [7]: %timeit per.month
```
Before:
```
The slowest run took 15.03 times longer than the fastest. This could mean that an intermediate result is being cached.
1000000 loops, best of 3: 1.4 µs per loop
```
After:
```
The slowest run took 25.64 times longer than the fastest. This could mean that an intermediate result is being cached.
1000000 loops, best of 3: 744 ns per loop
```

- [ ] closes #xxxx
- [ ] tests added / passed
- [ ] passes `git diff upstream/master -u -- "*.py" | flake8 --diff`
- [ ] whatsnew entry

### CODE REVIEW

File: pandas/_libs/tslibs/frequencies.pyx [hunk 1]
```diff
@@ -0,0 +1,204 @@
...
+cimport cython
+
+import numpy as np
+cimport numpy as np
+np.import_array()  # Without this, `is_integer_object` causes segfaults
+
+from util cimport is_integer_object
+
+
+@cython.ccall
```
Comment (jreback) [inline pandas/_libs/tslibs/frequencies.pyx]: what does this decorator do?

**Review by jreback — CHANGES_REQUESTED** [review]
if u are claiming. perf boost then pls add asv

File: pandas/_libs/tslibs/frequencies.pyx (hunk 1 above)
Comment (jbrockmendel) [reply pandas/_libs/tslibs/frequencies.pyx]:
```
@cython.ccall
def foo(x):
```

is equivalent to `cpdef foo(x):`

**Comment by jbrockmendel** [comment]
asv is throwing build-time errors.  I've filed an issue there.  For now the timeit results will have to do.

**Review by jreback — CHANGES_REQUESTED** [review]
i still want actual asv code

**Comment by jbrockmendel** [comment]
@jorisvandenbossche Any chance you want to play "good cop" here?  I could use some help with asv, see error discussion [here](https://github.com/pandas-dev/pandas/pull/17274#issuecomment-326675088).

**Comment by jreback** [comment]
@jbrockmendel you need to add asv benchmarks to he code you are pushing 

you are touching highly perf sensistive code and it is very easy to break things
think of these like tests

**Comment by jbrockmendel** [comment]
Just pushed a commit that adds asv checks for Period property lookups.  I can't get asv to run it, would appreciate it if someone else would give it a shot.

File: pandas/_libs/tslibs/frequencies.pyx (hunk 1 above)
Comment (jreback) [reply pandas/_libs/tslibs/frequencies.pyx]: then use ``cpdef foo(x)``; don't try to change the existing style in a single PR.

**Review by jreback — CHANGES_REQUESTED** [review]

**Comment by jbrockmendel** [comment]
> then use cpdef foo(x); don't try to change the existing style in a single PR.

Note that this code was moved from non-cython, so is not changing existing style.  The idea was that where possible, retaining pure-python compatibility would be nice.  But not a big deal, so will change.

**Comment by jbrockmendel** [comment]
Looks like Travis error is an unrelated HTTP timeout.

File: pandas/_libs/tslibs/frequencies.pyx [hunk 2]
```diff
@@ -0,0 +1,201 @@
+# -*- coding: utf-8 -*-
+# cython: profile=False
+import re
+
+cimport cython
+
+import numpy as np
+cimport numpy as np
+np.import_array()  # Without this, `is_integer_object` causes segfaults
```
Comment (jreback) [inline pandas/_libs/tslibs/frequencies.pyx]: remove the comment, this is standard to do.

**Comment by jreback** [comment]
ok this looks good. just remove that comment and push again to get everything passing. ping on green.

**Review by jreback — APPROVED** [review]

**Comment by jbrockmendel** [comment]
ping.

**Comment by jreback** [comment]
https://travis-ci.org/pandas-dev/pandas/jobs/272723119 is failing. FYI just because they are in the allowed failing section doesn't mean they actually should fail. this is the build test, which installs like a user, w/o a development build.

File: setup.py [hunk 3]
```diff
@@ -492,6 +493,8 @@ def pxd(name):
                      'sources': ['pandas/_libs/src/datetime/np_datetime.c',
                                  'pandas/_libs/src/datetime/np_datetime_strings.c',
                                  'pandas/_libs/src/period_helper.c']},
+    '_libs.tslibs.frequencies': {'pyxfile': '_libs/tslibs/frequencies',
```
Comment (jreback) [inline setup.py]: this *also* needs to be listed in the ``setup`` function, IOW all sub-dirs have to be listed, put it after ``pandas._libs``

**Review by jreback — CHANGES_REQUESTED** [review]
see comments

**Comment by jbrockmendel** [comment]
Looks like the allowed-fail mentioned earlier is now OK, but the 3.5 OSX is now failing, says it just stalled and received no output for 10 minutes.  The only change since last night's all-green is the entry in the setup.py packages.

**Review by jreback — APPROVED** [review]

**Pull request merged**